torch/lib/c10d/TCPStore.cpp

```diff
@@ -769,22 +769,61 @@ class TCPServer {
  private:
   std::uint16_t port_;
   std::unique_ptr<TCPStoreMasterDaemon> daemon_;
+
+  // We store weak references to all TCPServers for which the caller requested
+  // multi-tenancy.
+  static std::unordered_map<std::uint16_t, std::weak_ptr<TCPServer>>
+      cachedServers_;
+
+  static std::mutex cache_mutex_;
 };
 
+std::unordered_map<std::uint16_t, std::weak_ptr<TCPServer>>
+    TCPServer::cachedServers_{};
+
+std::mutex TCPServer::cache_mutex_{};
+
 std::shared_ptr<TCPServer> TCPServer::start(const TCPStoreOptions& opts) {
+  auto startCore = [&opts]() {
+    TCPSocket socket{};
+    std::uint16_t port{};
+
+    std::tie(socket, port) = tcputil::listen(opts.port);
+
+    auto daemon = std::make_unique<TCPStoreMasterDaemon>(std::move(socket));
+
+    return std::make_shared<TCPServer>(port, std::move(daemon));
+  };
+
+  std::shared_ptr<TCPServer> server{};
+
   if (opts.multiTenant) {
-    LOG(WARNING)
-        << "The multi-tenant feature of TCPStore is not implemented yet.";
-  }
+    std::lock_guard<std::mutex> guard{cache_mutex_};
+
+    // If the caller is okay with a multi-tenant store, first check if we already
+    // have a TCPServer running on the specified port.
+    if (opts.port > 0) {
+      auto pos = cachedServers_.find(opts.port);
+      if (pos != cachedServers_.end()) {
+        server = pos->second.lock();
+        if (server != nullptr) {
+          return server;
+        }
 
-  TCPSocket socket{};
-  std::uint16_t port{};
+        // Looks like the TCPStore has been disposed, make sure that we release
+        // the control block.
+        cachedServers_.erase(pos);
+      }
+    }
 
-  std::tie(socket, port) = tcputil::listen(opts.port);
+    server = startCore();
 
-  auto daemon = std::make_unique<TCPStoreMasterDaemon>(std::move(socket));
+    cachedServers_.emplace(server->port(), server);
```

Comment:
Do we need this to be a map instead of a single cached server because we want to support `TCPStore::start()` with multiple ports? Curious if we use that feature anywhere or if there are any use cases for it.

Reply:
Yes, we need a map because there can be multiple `TCPStore` instances using different ports. For instance:

```python
s1 = TCPStore("localhost", port=1234, multi_tenant=True)
s2 = TCPStore("localhost", port=1234, multi_tenant=True)

s3 = TCPStore("localhost", port=5678, multi_tenant=True)
s4 = TCPStore("localhost", port=5678, multi_tenant=True)
```

Besides being pedantically correct (since we are offering an API and should not make any assumptions about how our types are going to be used), one use case that comes to my mind where this can be useful is TorchElastic. There we can use two multi-tenant `TCPStore` instances, one for the rendezvous itself, and one for the worker. This way we can avoid unnecessarily allocating a second port.